### ISSUE TITLE
Bug: Parser: Tokens for `let`, `static`, `yield` as identifiers differ from Espree

### ISSUE THREAD

**Issue by overlookmotel** [typescript-eslint/typescript-eslint#11989]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&sourceType=script&showAST=es&fileType=.js&code=MYewdgzgLgBCBGArGBeGBvAUDGB6XMAkgCYCmYUAlgGaWkBOMUIA1uRNjNSCAFxc8ANJwCGEAJ5hg-MZODCc%2BGAGlS4gO4h6xJq3acANqSj8jUBTGgiq0y1GuV5ncXQPF%2BL0m%2BEBfANyYQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```TypeScript
// script
const obj = {
  let: let,
  static: static,
  yield: yield,
};
```

### ESLint Config

```javascript
```

### tsconfig

```jsonc
```

### Expected Result

I am actually not sure if this should be considered a bug or not. But it does not match Espree (ESLint).

Surprisingly, Espree produces `Keyword` tokens for `let`, `static` and `yield` in identifier positions, rather than `Identifier` tokens. [AST explorer](https://ast-explorer.dev/#eNpFjkEOAiEMRa9Cup4TYLyFccUGsSaMSAntmEwmc3dLWbjpz/s/7e8BBTys8Rs59dwEFmhqILeOqJAUElUWR4/VXd0RqnMvIj/GMiDyXpOfYkZB8WMYsETJGk81a89Ynn6KGuclVC0iLbLjAZi2nvC2NwzgB9tnAWw7AKZPvGPnTHXmJQryPxd6Y+URSd9QC+D8AaBbS2c=)

Personally, I think it is questionable if this is ideal behavior, but as it matches Esprima, I imagine ESLint's team will be very unlikely to agree to change it.

Some ESLint rules e.g. `no-extra-parens` depend on this. Here's one of the test cases for that rule:

```js
// script
(let)[foo]
```

With Espree parser, this produces no errors, but with TS-ESLint parser it flags an error incorrectly.

[ESLint playground](https://eslint.org/play/#eyJ0ZXh0IjoiKGxldClbZm9vXSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tbWFnaWMtbnVtYmVycyI6WyJlcnJvciJdLCJuby1leHRyYS1wYXJlbnMiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQifX19)

### Actual Result

Tokens for `let`, `static` and `yield` are `type: "Identifier"`, same as `foo` in this example:

```js
const obj = {
  foo: foo,
};
```

### Additional Info

I have no particular opinion on whether this should be changed or not.

* Pro: Align with Espree/ESLint
* Con: TS-ESLint's current behavior makes more sense IMO

I've opened this issue primarily to get clarity on whether this is considered a bug or not.

Either way, it's *mostly* an edge case with little practical impact. No-one names variables `let`. Though `static` might be more common.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `8.53.1` |

**Comment by bradzacher** [comment]
> Surprisingly, Espree produces Keyword tokens for let, static and yield in identifier positions, rather than Identifier tokens

I could try to dig up the issue - but essentially this is expected for them. They don't have any logic to handle contextual keywords - so their parser always emits it as a keyword token.

TypeScript on the other hand does properly handle things and hence we emit the correct tokens. I believe babel also emits the right tokens?

I raised it with espree and they said that it wasn't going to change. And I ultimately decided that we were right and it wasn't worth us trying to align with them because (a) they're wrong and (b) it's so so so rare that rules do actually care about the exact type of these tokens for it to actually matter.

In this case I would file a bug with eslint upstream about the broken rule. They are aiming for support of our AST in core rules and so they should handle it correctly.

**Comment by overlookmotel** [comment]
Fair enough! I agree that TS-ESLint's behavior is more correct / makes more sense. Sorry that I didn't manage to dig up the previous conversation before opening this issues.
